applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_nodal_reaction_response_function.cpp

```diff
@@ -0,0 +1,347 @@
+//       _____/ \__|_|   \__,_|\___|\__|\__,_|_|  \__,_|_| MECHANICS
+//
+//  License:		 BSD License
+//					 license: structural_mechanics_application/license.txt
+//
+//  Main authors:    Martin Fusseder, https://github.com/MFusseder
+//
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "adjoint_nodal_reaction_response_function.h"
+#include "utilities/variable_utils.h"
+#include "processes/find_nodal_neighbours_process.h"
+#include "processes/find_conditions_neighbours_process.h"
+
+namespace Kratos
+{
+    AdjointNodalReactionResponseFunction::AdjointNodalReactionResponseFunction(ModelPart& rModelPart, Parameters ResponseSettings)
+    : AdjointStructuralResponseFunction(rModelPart, ResponseSettings)
+    {
+        // Get id of node where a reaction should be traced
+        const int id_traced_node = ResponseSettings["traced_node_id"].GetInt();
+
+        // Get the corresponding dof to the reaction which should be traced
+        // by this response function e.g. REACTION_X, REACTION_MOMENT_X,...
+        mTracedReactionLabel = ResponseSettings["traced_reaction"].GetString();
+        // Get the corresponding dof to the displacement e.g. REACTION_X --> DISPLACEMENT_X
+        mTracedDisplacementDofLabel = this->GetCorrespondingDisplacementDofLabel(mTracedReactionLabel);
+
+        if(ResponseSettings.Has("adjust_influence_function"))
+            mAdjustInfluenceFunction = ResponseSettings["adjust_influence_function"].GetBool();
+
+        // Get pointer to traced node
+        mpTracedNode = rModelPart.pGetNode(id_traced_node);
+
+        // Check the given variables after reading
+        this->PerformResponseVariablesCheck();
+
+        // Find neighbour elements and conditions because they are needed to construct the partial derivatives
+        FindNodalNeighboursProcess neigbhor_elements_finder = FindNodalNeighboursProcess(mrModelPart, 10, 10);
+        FindConditionsNeighboursProcess neigbhor_conditions_finder = FindConditionsNeighboursProcess(mrModelPart, 10, 10);
+        neigbhor_elements_finder.Execute();
+        neigbhor_conditions_finder.Execute();
+        mpNeighborElements = mpTracedNode->GetValue(NEIGHBOUR_ELEMENTS);
+        mpNeighborConditions = mpTracedNode->GetValue(NEIGHBOUR_CONDITIONS);
+    }
+
+    AdjointNodalReactionResponseFunction::~AdjointNodalReactionResponseFunction(){}
+
+    void AdjointNodalReactionResponseFunction::InitializeSolutionStep()
+    {
+        KRATOS_TRY;
+
+        const VariableComponentType& r_traced_adjoint_dof =
+            KratosComponents<VariableComponentType>::Get(std::string("ADJOINT_") + mTracedDisplacementDofLabel);
+
+        // check if the given node is really fixed in the direction of the given dof.
+        if((mpTracedNode->GetDof(r_traced_adjoint_dof)).IsFree())
+            KRATOS_ERROR << "Chosen reaction '" << mTracedReactionLabel << "' is not fixed on node #" << mpTracedNode->Id() << "!" << std::endl;
```

Comment:
The dof r_traced_adjoint_dof corresponding to the chosen reaction mTracedReactionLabel ...